wkcuber/export_wkw_as_tiff.py

```diff
@@ -58,56 +78,95 @@ def wkw_name_and_bbox_to_tiff_name(name: str, slice_index: int) -> str:
+    else:
+        # swap axis and move the channel axis
+        data_slice = data_slice.transpose((2, 1, 0))
+    return Image.fromarray(data_slice)
 
 
 def export_tiff_slice(
-    export_args: Tuple[int, Tuple[Dict[str, Tuple[int, int, int]], str, str, str]]
+    export_args: Tuple[
+        int,
+        Tuple[
+            Dict[str, Tuple[int, int, int]], str, str, str, Union[None, Tuple[int, int]]
+        ],
+    ]
 ):
-    slice_number, (bbox, dest_path, name, dataset_path) = export_args
-    tiff_bbox = bbox
+    slice_number, (tiff_bbox, dest_path, name, dataset_path, tiling_size) = export_args
     tiff_bbox["topleft"] = [
         tiff_bbox["topleft"][0],
         tiff_bbox["topleft"][1],
         tiff_bbox["topleft"][2] + slice_number,
     ]
     tiff_bbox["size"] = [tiff_bbox["size"][0], tiff_bbox["size"][1], 1]
 
-    tiff_file_name = wkw_name_and_bbox_to_tiff_name(name, slice_number)
+    if tiling_size is None:
+        with wkw.Dataset.open(dataset_path) as dataset:
+            tiff_data = dataset.read(tiff_bbox["topleft"], tiff_bbox["size"])
+        tiff_file_name = wkw_name_and_bbox_to_tiff_name(name, slice_number)
 
-    tiff_file_path = os.path.join(dest_path, tiff_file_name)
+        tiff_file_path = os.path.join(dest_path, tiff_file_name)
 
-    with wkw.Dataset.open(dataset_path) as dataset:
-        tiff_data = dataset.read(tiff_bbox["topleft"], tiff_bbox["size"])
+        image = wkw_slice_to_image(tiff_data)
+        image.save(tiff_file_path)
+        logging.info(f"saved slice {slice_number}")
 
-    # discard the z dimension
-    tiff_data = tiff_data.squeeze(axis=3)
-    if tiff_data.shape[0] == 1:
-        # discard greyscale dimension
-        tiff_data = tiff_data.squeeze(axis=0)
-        # swap the axis
-        tiff_data = tiff_data.transpose((1, 0))
     else:
-        # swap axis and move the channel axis
-        tiff_data = tiff_data.transpose((2, 1, 0))
+        tile_bbox = copy.deepcopy(tiff_bbox)
+        tile_bbox["size"] = [tiling_size[0], tiling_size[1], 1]
+        with wkw.Dataset.open(dataset_path) as dataset:
+
+            for y_tile_index in range(
+                1, ceil((tiff_bbox["size"][1]) / tiling_size[1]) + 1
+            ):
+                tile_bbox["topleft"][1] += tiling_size[1]
+                tile_tiff_path = os.path.join(
+                    dest_path, str(slice_number), str(y_tile_index)
+                )
+                os.makedirs(tile_tiff_path, exist_ok=True)
 
-    logging.info(f"saving slice {slice_number}")
+                for x_tile_index in range(
+                    1, ceil((tiff_bbox["size"][0]) / tiling_size[0]) + 1
+                ):
+                    tile_tiff_filename = f"{x_tile_index}.tiff"
+                    tile_bbox["topleft"][0] += tiling_size[0]
 
-    image = Image.fromarray(tiff_data)
-    image.save(tiff_file_path)
+                    tile_tiff_data = dataset.read(
+                        tile_bbox["topleft"], tile_bbox["size"]
```

Comment:
The handling of the `tile_bbox` can be improved a bit, I think :) Mainly, I'd avoid mutating the bounding box variable, since it makes deep copy and the restoration necessary (additionally, it scatters the bbox computation across the entire loops, even though it could be done in one statement). Instead, I'd compute the `tile_bbox` from scratch where it's necessary (line 143 probably) using the `*_tile_index` variables (which I would make zero-based so that the range expressions are easier to read).

Roughly, it would look like this:
```python
tile_bbox_size = [tiling_size[0], tiling_size[1], 1]
for y_tile_index in range(ceil((tiff_bbox["size"][1]) / tiling_size[1])):
  tile_tiff_path = os.path.join(
    dest_path, str(slice_number), str(y_tile_index + 1)
  )
  os.makedirs...
  for x_tile_index in range(ceil((tiff_bbox["size"][0]) / tiling_size[0])):
    tile_tiff_filename = f"{x_tile_index + 1}.tiff"
    tile_bbox_topleft = [tiff_bbox[0] + tile_bbox_size[0] * x_tile_index, ...]
    ...
```

Reply:
Ok I have no excuse here why I did it that way but yes your way is definitely better than mine ans is way more readable. Thx :D